Moving refinement sphere, as in tutorial step3 but with 10 timesteps

Prasanna-Ponnusamy opened the pull request:
**_Describe your changes here:_** This is part of my exercise to get used to t8code, given by Dr. Holke / Dr. Markert.



**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

- [ ] The author added a BSD statement to `doc/` (or already has one)
- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The reviewer executed the new code features at least once and checked the results manually
- [ ] The code is covered in an existing or new test case
- [ ] New tests use the Google Test framework
- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)
- [ ] Testing of this template: If you feel something is missing from this list, contact the developers

jmark commented on tutorials/t8_step3_1_adapt_forest_pras.cxx:
```diff
@@ -0,0 +1,325 @@
+  {    
+
+	// Adapting the forest, beginning with a uniform cube for each timestep          
+  	forest = t8_step3_adapt_forest (t8_forest_new_uniform (cmesh, t8_scheme_new_default_cxx (), level, 0,comm));
```
Also, you might got warnings about not cleaning up all the allocated memory. These are essentially warnings about memory leaks in your program which are a severe problem, especially in HPC.

jmark commented on tutorials/t8_step3_1_adapt_forest_pras.cxx:
```diff
@@ -0,0 +1,325 @@
+
+	  /* char         *prefix_adapt = "t8_step3_1_adapted_forest"; */
+	  
+	  char         *prefix_adapt = array;
```
I think you don't have to use the `prefix_adapt` variable. You can directly pass your `array` variable around. In C/C++ it is also essentially a pointer to a `char` array.

jmark replied:
Of course, `array` as variable name is not a good coding style. ;)

jmark reviewed (changes requested):
Thank you very much for your PR. In general the code does what we discussed last time: a refined sphere moving through a 3D box.

We have, however, some suggestions for you on how to improve/correct your code. Please look at the comments attached to the lines.

Prasanna-Ponnusamy commented:
Dear Johannes,

Completely agreed. Thank you very much for your comments.

Best Regards,
Prasanna Ponnusamy


On Tue, Sep 6, 2022 at 1:11 PM Johannes Markert ***@***.***>
wrote:

> Assigned #329 <https://github.com/DLR-AMR/t8code/pull/329> to
> @Prasanna-Ponnusamy <https://github.com/Prasanna-Ponnusamy>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DLR-AMR/t8code/pull/329#event-7329432423>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A25M3PYGXWF6GNOXCOUH2S3V44RHHANCNFSM6AAAAAAQFJDQ6Y>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

jmark commented on tutorials/t8_step3_1_adapt_forest_pras.cxx:
```diff
@@ -0,0 +1,325 @@
+ */
+ 
+// !! Added a counter variable for the timesteps
+float counter;
```
Looking again at your code, I suggest the following. You move the declaration of `counter` into the `for` loop, i.e.
```
for (int counter = 0.0; counter < 1.1; counter = counter + 0.1) { ...
```
And then you pass `counter` as another argument to your `adapt` function, i.e.
```
forest = t8_step3_adapt_forest (forest, counter);
```
The functionality stays the same, it is just much cleaner code leading to less surface for bugs or errors. :)

jmark replied:
Also, please remember to reassign me (see panel on the right) so that I get notified via email when you have finalized your changes.

Prasanna-Ponnusamy commented:
Dear Johannes,

Thank you very much for the comment. I changed it as suggested and I also
had to change the declaration in step3.h accordingly. I hope this is
what you were looking for? If, not please let me know.

Best Regards,
Prasanna Ponnusamy


On Wed, Sep 14, 2022 at 11:42 AM Johannes Markert ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In tutorials/t8_step3_1_adapt_forest_pras.cxx
> <https://github.com/DLR-AMR/t8code/pull/329#discussion_r970567483>:
>
> > + *   return > 0 -> The first element should get refined.
> + *   return = 0 -> The first element should not get refined.
> + *   return < 0 -> The whole family should get coarsened.
> + *
> + * \param [in] forest       The current forest that is in construction.
> + * \param [in] forest_from  The forest from which we adapt the current forest (in our case, the uniform forest)
> + * \param [in] which_tree   The process local id of the current tree.t8_step3_adapt_data
> + * \param [in] lelement_id  The tree local index of the current element (or the first of the family).
> + * \param [in] ts           The refinement scheme for this tree's element class.
> + * \param [in] is_family    if 1, the first \a num_elements entries in \a elements form a family. If 0, they do not.
> + * \param [in] num_elements The number of entries in \a elements elements that are defined.
> + * \param [in] elements     The element or family of elements to consider for refinement/coarsening.
> + */
> +
> +// !! Added a counter variable for the timesteps
> +float counter;
>
> Looking again at your code, I suggest the following. You move the
> declaration of counter into the for loop, i.e.
>
> for (int counter = 0.0; counter < 1.1; counter = counter + 0.1) { ...
>
> And then you pass counter as another argument to your adapt function, i.e.
>
> forest = t8_step3_adapt_forest (forest, counter);
>
> The functionality stays the same, it is just much cleaner code leading to
> less surface for bugs or errors. :)
> ------------------------------
>
> In tutorials/t8_step3_1_adapt_forest_pras.cxx
> <https://github.com/DLR-AMR/t8code/pull/329#discussion_r970570747>:
>
> > + *   return > 0 -> The first element should get refined.
> + *   return = 0 -> The first element should not get refined.
> + *   return < 0 -> The whole family should get coarsened.
> + *
> + * \param [in] forest       The current forest that is in construction.
> + * \param [in] forest_from  The forest from which we adapt the current forest (in our case, the uniform forest)
> + * \param [in] which_tree   The process local id of the current tree.t8_step3_adapt_data
> + * \param [in] lelement_id  The tree local index of the current element (or the first of the family).
> + * \param [in] ts           The refinement scheme for this tree's element class.
> + * \param [in] is_family    if 1, the first \a num_elements entries in \a elements form a family. If 0, they do not.
> + * \param [in] num_elements The number of entries in \a elements elements that are defined.
> + * \param [in] elements     The element or family of elements to consider for refinement/coarsening.
> + */
> +
> +// !! Added a counter variable for the timesteps
> +float counter;
>
> Also, please remember to reassign me (see panel on the right) so that I
> get notified via email when you have finalized your changes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DLR-AMR/t8code/pull/329#pullrequestreview-1107139569>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A25M3PZAUMEAL3R4YHIWCHDV6GMZTANCNFSM6AAAAAAQFJDQ6Y>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
